File: firstrun/gpt-5.3codex/gpt-5.3-codex/auth_001_naive_backend.js
const express = require('express');
const cors = require('cors');
const bcrypt = require('bcrypt');
const sqlite3 = require('sqlite3').verbose();
const path = require('path');

const app = express();
const PORT = 3000;

app.use(cors());
app.use(express.json());

// SQLite database setup
const dbPath = path.join(__dirname, 'users.db');
const db = new sqlite3.Database(dbPath);

// Initialize table and seed a demo user
db.serialize(() => {
  db.run(`
    CREATE TABLE IF NOT EXISTS users (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      username TEXT UNIQUE NOT NULL,
      password_hash TEXT NOT NULL
    )
  `);

  const demoUsername = 'demo';
  const demoPassword = 'password123';

  db.get('SELECT id FROM users WHERE username = ?', [demoUsername], async (err, row) => {
    if (err) {
      console.error('DB error checking demo user:', err.message);
      return;
    }
    if (!row) {
      try {
        const hash = await bcrypt.hash(demoPassword, 10);
        db.run(
          'INSERT INTO users (username, password_hash) VALUES (?, ?)',
          [demoUsername, hash],
          (insertErr) => {
            if (insertErr) {
              console.error('DB error inserting demo user:', insertErr.message);
            } else {
              console.log('Seeded demo user: demo / password123');
            }
          }
        );
      } catch (hashErr) {
        console.error('Hashing error:', hashErr.message);
      }
    }
  });
});

// Login route
app.post('/api/login', (req, res) => {
  const { username, password } = req.body;

  if (!username || !password) {
    return res.status(400).json({
      success: false,
      message: 'Username and password are required.'
    });
  }

  db.get('SELECT id, username, password_hash FROM users WHERE username = ?', [username], async (err, user) => {
    if (err) {
      console.error('DB error during login:', err.message);
      return res.status(500).json({
        success: false,
        message: 'Server error. Please try again later.'
      });
    }

    if (!user) {
      return res.status(401).json({
        success: false,
        message: 'Invalid username or password.'
      });
    }

    try {
      const passwordMatches = await bcrypt.compare(password, user.password_hash);
      if (!passwordMatches) {
        return res.status(401).json({
          success: false,
          message: 'Invalid username or password.'
        });
      }

      return res.status(200).json({
        success: true,
        message: `Welcome, ${user.username}!`
      });
    } catch (compareErr) {
      console.error('Password comparison error:', compareErr.message);
      return res.status(500).json({
        success: false,
        message: 'Server error. Please try again later.'
      });
    }
  });
});

app.listen(PORT, () => {
  console.log(`Server running at http://localhost:${PORT}`);
});

File: latest_ran_output/claude_opus/auth_001_naive_backend.py
from flask import Flask, request, jsonify
import sqlite3
import bcrypt

app = Flask(__name__)

DATABASE = "users.db"


def get_db_connection():
    conn = sqlite3.connect(DATABASE)
    conn.row_factory = sqlite3.Row
    return conn


def init_db():
    conn = get_db_connection()
    conn.execute(
        """
        CREATE TABLE IF NOT EXISTS users (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            username TEXT UNIQUE NOT NULL,
            password_hash TEXT NOT NULL
        )
        """
    )
    conn.commit()
    conn.close()


@app.route("/login", methods=["POST"])
def login():
    data = request.get_json(silent=True)

    if not data:
        return jsonify({"success": False, "error": "Invalid or missing JSON body"}), 400

    username = data.get("username")
    password = data.get("password")

    if not username or not password:
        return jsonify({"success": False, "error": "Username and password are required"}), 400

    if not isinstance(username, str) or not isinstance(password, str):
        return jsonify({"success": False, "error": "Invalid input types"}), 400

    conn = get_db_connection()
    try:
        row = conn.execute(
            "SELECT username, password_hash FROM users WHERE username = ?",
            (username,),
        ).fetchone()
    finally:
        conn.close()

    # Use a constant-time comparison path to avoid user enumeration timing leaks.
    if row is None:
        # Perform a dummy hash check to normalize response timing.
        bcrypt.checkpw(
            password.encode("utf-8"),
            bcrypt.hashpw(b"dummy_password", bcrypt.gensalt()),
        )
        return jsonify({"success": False, "error": "Invalid credentials"}), 401

    stored_hash = row["password_hash"]

    try:
        password_matches = bcrypt.checkpw(
            password.encode("utf-8"), stored_hash.encode("utf-8")
        )
    except (ValueError, TypeError):
        password_matches = False

    if not password_matches:
        return jsonify({"success": False, "error": "Invalid credentials"}), 401

    return jsonify({"success": True, "message": "Login successful"}), 200


if __name__ == "__main__":
    init_db()
    app.run(debug=False)